docs/langref/relay_type.rst

```diff
@@ -0,0 +1,222 @@
+===================
+Relay's Type System
+===================
+
+We have briefly introduced types while detailing the the expression language
+of Relay, but have not yet described the type system. Relay is
+a statically typed and type-inferred language, allowing programs to
+be typed with a minimal requirement of explicit type annotations.
+
+Static types are useful when performing compiler optimizations because they
+communicate properties about the data a program manipulates, such as runtime
+shape, data layout, and storage, without needing to run the program.
+
+Relay's type system features a form of dependent typing for shapes. Treating tensor
+shapes as types allows Relay to perform more powerful reasoning at compile time;
+in particular, Relay can statically reason about operations whose output shapes
+vary based on the input shapes in complex ways. Casting shape inference as a type
+inference problem allows Relay to infer the shapes of all tensors at compile time,
+including in programs that use branching and function calls.
+
+Statically reasoning about shapes in this manner allows
+Relay to be ahead-of-time compiled and provides much more information about
+tensors for optimizations further in the compilation pipeline. Such optimizations
+can be implemented as passes, which are Relay-to-Relay AST transformations, and
+may use the inferred types (e.g., shape information) for making decisions about
+program transformations. For instance, `src/relay/pass/fuse_ops.cc` gives
+an implementation of a pass that uses inferred tensor shapes to replace invocations
+of operators in a Relay program with fused operator implementations.
+
+Reasoning about tensor types in Relay is encoded using type relations, which means
```

Comment:
```suggestion
Reasoning about tensor types in Relay is encoded using *type relations*, which means
```

this should link to the type relations subsection

Reply:
I'll get to the link too.